keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1121 @@
+
+The exact permissions are going to be a shared responsibility between extension
+authors and in-tree Kubernetes developers, since anyone can author a CRD.
+
+As part of this KEP, we will update the schemas to supply permission verbs for
+the fields listed above. We will document the patterns used so that extension
+authors can make their own permission verbs which won't conflict with each other
+or with in-tree verbs. We may not know exactly what the best practices are until
+we have an alpha or beta behind us, hence we're not ready to specify that in
+this KEP yet.
+
+(The exact permission system will be described below because we're not done with
+the requirements yet, keep reading!)
+
+#### API Server Additional Plumbing and Checks
+
+Currently apiserver performs the authz check before descending into any detailed
+logic for mutating operations. This means that when it is doing the authz check,
+it knows the *object* that is being changed, but nothing about *which fields*
+are changing.
+
+We will introduce a new permission, "granular". This permission
+means that although the caller doesn't have generic update permissions, it may
+still change some fields if it has the specific permissions required.
+
+API Server already maintains schema information for all resource types. This
+will be watched to also maintain field <-> permission mappings.
+
+Now we can describe the modifications to API Server:
+
+On any mutating call,
+
+1. Check the "PUT" / "PATCH" / "CREATE" permission. If the actor has this,
+   proceed to step 3.
+2. Otherwise, check the "granular" permission. If yes, place a
+   marker in the request context and proceed to step 3, otherwise, fail the
+   request with a forbidden error.
+3. Compute the change (patch logic, SSA logic, defaulting etc). Compute a list
+   of fields which changed. (SSA logic makes this easy.)
+4. Pre- or post- (DECISION NEEDED) webhooks, check the list of fields:
+5. If the marker is present, every field needs to be covered by some permission.
+   Check permissions, stop and fail if some field is not covered. (The order is
+   described below.)
+6. Otherwise (no marker) ONLY fields having "excluded" permissions need to be
+   checked; if such fields are modified, check their permissions as in step 5.
+7. If all checks pass, perform the rest of the needed operation.
+8. If any check fails, fail the request. The error message will not list ALL
+   permissions that are lacked, because it is possible to craft a request that
+   requires a large number of authz requests.
+
+When fine-grained field permissions need to be checked, they will be checked
+in the order of most general to most specific. When we describe the
+representation in the schema it will be clear what that means. The reason for
+this is that multiple fields can be covered by a more general permission,
+greatly reducing the number of checks needed in the worst case (as long as the
```

Comment:
Should we expand the authorization interface to have a slice option to pass more than one to evaluate a time?  We could expand the webhook using another version and non-participating webhooks could be shimmed to answer one at a time in parallel

Reply:
that's funny, I literally had that thought this morning

Reply:
that would already be nice for impersonation

Reply:
I have no objection to that but it sounds like an optimization that can be done whether or not the KEP goes through.

Reply:
> I have no objection to that but it sounds like an optimization that can be done whether or not the KEP goes through.

I'm not completely sure of that.  There are two "obvious" ways to handle parameterization of something like label values.  

1. One way is to describe in the schema a manner in which shrink the keyspace (you've done this).
2.  Another way is to send all the label keys to the authorizer as distinct verbs and let the authorizer decide how best to let a user/cluster-admin express rules.

I think if we have a bulk authorization interface, it becomes more practical to lookup every key at once.

Reply:
Sorry it says you sent this 7 days ago but this is the first I saw it -- if you look at my comment [here](https://github.com/lavalamp/enhancements/pull/7#discussion_r1001229920) I think my current plan doesn't require a bulk lookup. My backup plan also doesn't require a bulk lookup.

Reply:
The latest revision doesn't require bulk lookup, either.